### PR TITLE
[Feature][linkis-engineplugin-spark] spark submit jar (engineConnMode is once)

### PR DESCRIPTION
**Pull request by rarexixi**
spark submit jar (engineConnMode is once)

### Related issues/PRs

Related issues: #3421

### Brief change log

- Add spark submit once feature.

### CODE REVIEW

**Review by Davidhua1996 — DISMISSED** [review]
LGTM.

**Comment by Davidhua1996** [comment]
The modification of EngineConnServer and EngineConnHook should request reviewers from @wushengyeyouya @peacewong

File: linkis-computation-governance/linkis-client/linkis-computation-client/src/test/java/org/apache/linkis/computation/client/SparkOnceJobTest.java [hunk 1]
```diff
@@ -0,0 +1,127 @@
...
+          } else {
+            break;
+          }
+        }
+        for (String log : logLines) {
+          System.out.println(log);
+        }
+        fromLine += logLines.size();
+      } catch (Exception e) {
+        logger.error("获取日志失败", e);
```
Comment (peacewong) [inline linkis-computation-governance/linkis-client/linkis-computation-client/src/test/java/org/apache/linkis/computation/client/SparkOnceJobTest.java:97]: Use English

File: linkis-computation-governance/linkis-engineconn/linkis-engineconn-common/src/main/scala/org/apache/linkis/engineconn/common/conf/EngineConnConf.scala [hunk 2]
```diff
@@ -35,6 +35,11 @@ object EngineConnConf {
     "org.apache.linkis.engineconn.computation.executor.hook.ComputationEngineConnHook"
   )
 
+  val ENGINE_CONN_ONCE_HOOKS = CommonVars(
+    "wds.linkis.engine.connector.once.hooks",
```
Comment (peacewong) [inline linkis-computation-governance/linkis-engineconn/linkis-engineconn-common/src/main/scala/org/apache/linkis/engineconn/common/conf/EngineConnConf.scala:39]: remove prefix "wds"

File: linkis-computation-governance/linkis-engineconn/linkis-engineconn-common/src/main/scala/org/apache/linkis/engineconn/common/hook/EngineConnHook.scala [hunk 3]
```diff
@@ -52,27 +52,33 @@ trait EngineConnHook {
...
-            Utils.tryAndWarn(Utils.getClassInstance[EngineConnHook](clazz))
-          }
-          .filter(_ != null)
-      } else {
-        Array.empty
-      }
+      engineConnHooks = hooks
+        .split(",")
+        .map(_.trim)
+        .filter(_.nonEmpty)
```
Comment (peacewong) [inline linkis-computation-governance/linkis-engineconn/linkis-engineconn-common/src/main/scala/org/apache/linkis/engineconn/common/hook/EngineConnHook.scala:68]: Use StringUtils.isNotBlank

File: linkis-computation-governance/linkis-engineconn/linkis-engineconn-plugin-core/src/main/scala/org/apache/linkis/manager/engineplugin/common/launch/process/ProcessEngineConnLaunchBuilder.scala [hunk 4]
```diff
@@ -52,9 +52,7 @@ trait ProcessEngineConnLaunchBuilder extends EngineConnLaunchBuilder {
       engineConnBuildRequest: EngineConnBuildRequest
   ): util.List[String] = new util.ArrayList[String]
 
-  override def buildEngineConn(
-      engineConnBuildRequest: EngineConnBuildRequest
-  ): EngineConnLaunchRequest = {
+  override def buildEngineConn(): EngineConnLaunchRequest = {
     implicit val engineConnBuildRequestImplicit: EngineConnBuildRequest = engineConnBuildRequest
```
Comment (peacewong) [inline linkis-computation-governance/linkis-engineconn/linkis-engineconn-plugin-core/src/main/scala/org/apache/linkis/manager/engineplugin/common/launch/process/ProcessEngineConnLaunchBuilder.scala:56]: implicit can it be removed？

File: linkis-engineconn-plugins/spark/src/main/java/org/apache/linkis/engineplugin/spark/exception/ExecutorInitException.java [hunk 5]
```diff
@@ -0,0 +1,43 @@
...
+ */
+
+package org.apache.linkis.engineplugin.spark.exception;
+
+import org.apache.linkis.common.exception.ErrorException;
+
+public class ExecutorInitException extends ErrorException {
+
+  public static final int ERROR_CODE = 16021;
+
```
Comment (peacewong) [inline linkis-engineconn-plugins/spark/src/main/java/org/apache/linkis/engineplugin/spark/exception/ExecutorInitException.java:25]: recommended to use SparkErrorCodeSummary

File: linkis-engineconn-plugins/spark/src/main/java/org/apache/linkis/engineplugin/spark/exception/JobExecutionException.java [hunk 6]
```diff
@@ -0,0 +1,36 @@
...
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.linkis.engineplugin.spark.exception;
+
+import org.apache.linkis.common.exception.ErrorException;
+
+public class JobExecutionException extends ErrorException {
```
Comment (peacewong) [inline linkis-engineconn-plugins/spark/src/main/java/org/apache/linkis/engineplugin/spark/exception/JobExecutionException.java:22]: recommended to use SparkErrorCodeSummary

File: linkis-engineconn-plugins/spark/src/main/scala/org/apache/linkis/engineplugin/spark/utils/SparkJobProgressUtil.scala [hunk 7]
```diff
@@ -0,0 +1,117 @@
...
+import org.apache.linkis.common.utils.{JsonUtils, Logging}
+import org.apache.linkis.protocol.engine.JobProgressInfo
+
+import org.apache.commons.lang3.StringUtils
+import org.apache.http.client.methods.HttpGet
+import org.apache.http.impl.client.HttpClients
+import org.apache.http.util.EntityUtils
+
+object SparkJobProgressUtil extends Logging {
+  val YARN_RESTFUL_URL: String = CommonVars[String]("yarn.restful.url", "").getValue
```
Comment (peacewong) [inline linkis-engineconn-plugins/spark/src/main/scala/org/apache/linkis/engineplugin/spark/utils/SparkJobProgressUtil.scala:30]: 1. should add linkis.spark.once prefix  
2. Is it possible to support passing in and obtaining parameters at the same time, through YARN_RESTFUL_URL.getValue(engineCreationContext.getOptions)

File: linkis-engineconn-plugins/spark/src/main/scala/org/apache/linkis/engineplugin/spark/factory/SparkEngineConnFactory.scala [hunk 8]
```diff
@@ -51,8 +59,72 @@ class SparkEngineConnFactory extends MultiExecutorEngineConnFactory with Logging
...
       engineCreationContext: EngineCreationContext
   ): Any = {
+    if (EngineConnServer.isOnceMode) {
+      createSparkEngineConnContext(engineCreationContext)
+    } else {
+      createSparkEngineSession(engineCreationContext)
+    }
+  }
+
+  def createSparkEngineConnContext(
```
Comment (peacewong) [inline linkis-engineconn-plugins/spark/src/main/scala/org/apache/linkis/engineplugin/spark/factory/SparkEngineConnFactory.scala:69]: should rename createSparkEngineConnContext to createSparkOnceEngineConnContext

File: linkis-engineconn-plugins/spark/src/main/scala/org/apache/linkis/engineplugin/spark/launch/SparkEngineConnLaunchBuilder.scala [hunk 9]
```diff
@@ -0,0 +1,178 @@
...
+        getMainClass,
+        getGcLogDir(engineConnBuildRequest),
+        getLogDir(engineConnBuildRequest)
+      )
+    }
+  }
+
+  def isOnceMode: Boolean = {
+    val engineConnMode = engineConnBuildRequest.labels.asScala.find(label =>
+      "engineConnMode".equals(label.getLabelKey)
```
Comment (peacewong) [inline linkis-engineconn-plugins/spark/src/main/scala/org/apache/linkis/engineplugin/spark/launch/SparkEngineConnLaunchBuilder.scala:69]: LabelUtil can add getEngineConnModeLabel method

File: linkis-engineconn-plugins/spark/src/main/scala/org/apache/linkis/engineplugin/spark/utils/SparkJobProgressUtil.scala [hunk 10]
```diff
@@ -0,0 +1,119 @@
...
+          sparkJobResult.get("numCompletedTasks").asInstanceOf[Integer]
+        )
+      )
+    }
+  }
+
+  def getSparkJobInfo(applicationId: String): Array[java.util.Map[String, Object]] =
+    if (StringUtils.isBlank(applicationId)) Array.empty
+    else {
+      val getAppUrl = s"$YARN_RESTFUL_URL/ws/v1/cluster/apps/$applicationId"
```
Comment (peacewong) [inline linkis-engineconn-plugins/spark/src/main/scala/org/apache/linkis/engineplugin/spark/utils/SparkJobProgressUtil.scala:76]: Is it possible to support passing in and obtaining parameters at the same time, through YARN_RESTFUL_URL.getValue(engineCreationContext.getOptions)

**Review by peacewong — APPROVED** [review]
LGTM.

**Pull request merged**